src/ir_Corona.cpp

```diff
@@ -0,0 +1,586 @@
+                  kCoronaAcSwingVToggleOffset);
+}
+
+/// Set the Timer time
+/// @param[in] section index of section, used for offset.
+/// @param[in] nr_of_mins Number of minutes to set the timer to.
+///   (non in range value is disable).
+///   Valid is from 1 minute to 12 hours
+void IRCoronaAc::_setTimer(const uint8_t section, const uint16_t nr_of_mins) {
+  // default to off
+  uint16_t hsecs = kCoronaAcTimerOff;
+  if (1 <= nr_of_mins && nr_of_mins <= kCoronaAcTimerMax)
+    hsecs = nr_of_mins * kCoronaAcTimerUnitsPerMin;
+
+  uint8_t pos = section * kCoronaAcSectionBytes;
+  // convert 16 bit value to separate 8 bit parts
+  remote_state[pos + kCoronaAcSectionData1Pos] = hsecs >> 8;
+  remote_state[pos + kCoronaAcSectionData0Pos] = hsecs;
+}
+
+/// Get the current Timer time
+/// @return The number of minutes it is set for. 0 means it's off.
+/// @note The A/C protocol supports 2 second increments
+uint16_t IRCoronaAc::_getTimer(const uint8_t section) {
+  uint8_t pos = section * kCoronaAcSectionBytes;
+  // combine separate 8 bit parts to 16 bit value
+  uint16_t hsecs = remote_state[pos + kCoronaAcSectionData1Pos] << 8 |
+                   remote_state[pos + kCoronaAcSectionData0Pos];
+
+  if (hsecs == kCoronaAcTimerOff)
+    return 0;
+
+  return hsecs / kCoronaAcTimerUnitsPerMin;
+}
+
+/// Get the current On Timer time
+/// @return The number of minutes it is set for. 0 means it's off.
+uint16_t IRCoronaAc::getOnTimer(void) {
+  return _getTimer(kCoronaAcOnTimerSection);
+}
+
+/// Set the On Timer time
+/// @param[in] nr_of_mins Number of minutes to set the timer to.
+///   (0 or kCoronaAcTimerOff is disable).
+void IRCoronaAc::setOnTimer(const uint16_t nr_of_mins) {
+  _setTimer(kCoronaAcOnTimerSection, nr_of_mins);
+  // if we set a timer value, clear the other timer
+  if (getOnTimer())
+    setOffTimer(kCoronaAcTimerOff);
+  updatePowerButton();
+}
+
+/// Get the current Off Timer time
+/// @return The number of minutes it is set for. 0 means it's off.
+uint16_t IRCoronaAc::getOffTimer(void) {
+  return _getTimer(kCoronaAcOffTimerSection);
+}
+
+/// Set the Off Timer time
+/// @param[in] nr_of_mins Number of minutes to set the timer to.
+///   (0 or kCoronaAcTimerOff is disable).
+void IRCoronaAc::setOffTimer(const uint16_t nr_of_mins) {
+  _setTimer(kCoronaAcOffTimerSection, nr_of_mins);
+  // if we set a timer value, clear the other timer
+  if (getOffTimer())
+    setOnTimer(kCoronaAcTimerOff);
+  updatePowerButton();
+}
+
+/// Convert the internal state into a human readable string.
+/// @return The current internal state expressed as a human readable String.
+String IRCoronaAc::toString(void) {
+  String result = "";
+  result.reserve(140);  // Reserve some heap for the string to reduce fragging.
+  result += addBoolToString(getPower(), kPowerStr, false);
+  result += addBoolToString(getPowerButton(), kPowerButtonStr);
+  result += addModeToString(getMode(), 0xFF, kCoronaAcModeCool,
```

Comment:
@crankyoldgit  Would it be a bad idea to change this from generating
`"Power: On, Power Button: On, Mode .."` / `"Power: On, Power Button: Off, Mode .."`
to instead generating:
`"Power: On, Power Button, Mode .."` / `"Power: On, Mode .."`
That is `Power Button` is only included if pressed

Reply:
hmm.
I think the original is better. i.e. listing either state explicitly for the setting.

About the only things "exclude by omission" are timers. Or if a state isn't included in the message. i.e. Some protocols don't send the entire state in a single message.

So, the guide-of-thumb is: If you can determine the "state" of a setting from a message, report that state. The exception(s) are timers. e.g. If all the timers are off, there is no need to always explicitly say "Off Timer: Off, On Timer: Off" explicitly. Just "Timers: Off" etc. Also, if a timer is disabled, you don't need to report what ever "time" the timer is set to in the `toString()` output.
e.g. "On Timer: Off" vs. "On Timer: Off, On Time: 01:00" (or what ever value is in the timer bits etc.

Reply:
Ok gotcha, sounds good.

Reply:
It's mostly so some poor soul can construct the exact same message using the class methods.